man/man8/builddbm.8
.TH BUILDDBM 8 "20 Januari 1997" "" "Cistron Radius Daemon"
.SH NAME
builddbm -- build DBM file from RADIUS users file.
.SH SYNOPSIS
.B builddbm
.SH DESCRIPTION
This utility is used to build a DBM file (in DBM or NDBM format) from
the radius \fIusers\fP file to save memory.
.PP
Usage of this utility is only encouraged if you don't use multiple
DEFAULT entries (or only at the end of the \fIusers\fP file) and
you have more then 500 entries in the \fIusers\fP file. Normally,
\fBradiusd\fP reads the \fIusers\fP file into memory at startup. With
very large files this may cost too many resources and it may then
be wiser to use a DBM file to save memory.

Be aware that with the fastusers module, one can get very good results
with no DBM file, so use of this program may be depreciated.
.SH SEE ALSO
radiusd(8), users(5rad).
.SH AUTHOR
Miquel van Smoorenburg, miquels@cistron.nl.


doc/ChangeLog
FreeRADIUS 3.2.11 Wed 03 Jun 2026 12:00:00 UTC urgency=low
	Configuration changes
	* PEAP / SoH functionality has been removed.  It was removed from
	  Windows over 10 years ago, and is no longer relevant.
	* Add example of using socat to test radclient + UDP to the server
	  over TLS.  See sites-available/tls, and look for "socat".
	* Add per-section "do_xlat" and "is_json" config items to rlm_rest.
	  see mods-available/rest for details.
	* Add notes on the security of CHAP-Password and MS-CHAP-*
	  If you send these attributes over RADIUS/UDP, then any observer
	  can crack those passwords in milliseconds on a laptop.
	* Removed all code and configuration related to trustrouter (abfab)
	  That project was marked as end of life a few years ago.

	Feature improvements
	* Implement "track_home_server" in home_server_pool configuration
	  See proxy.conf.  The previous (and not working) "affinity_id"
	  configuration has been removed.
	* Implement EAP-PSK.  See mods-available/eap for configuration.
	  The PSK must be in "PSK-Password", and must be 16 octets in length
	* Add $VALUE{filename} which reads a value from one line of a
	  file.
	* Add $FILE{filename} which reads a value from the entire contents
	  of a file.
	* Update DPSK module to read VLANs from the CSV file, and to export
	  &reply:Pairwise-Master-Key
	* Clean up and normalize log messages and debug output.  These
	  changes make the messages easier to read, and easier to understand.
	* Add significantly more stats for outgoing proxy sockets.

	Bug fixes
	* Skip TLS PSK checks when OpenSSL does not support PSK.
	  Fixes #5874.
	* Updated to help build with OpenSSL 4.0 alpha.
	* Improve performance of State handling when tracking many EAP packets.
	* Fix overflow on relayed DHCP packets.  Issue found by A. Ramos.
	* Fix more corner cases in RadSec when sockets have errors.
	* Tweaked module messages to be more consistent, so that messages
	  associated with a request are printed with the correct request ID.
	* Fix race conditions when adding or deleting multiple dynamic TLS clients.
	* Ethernet addresses which are cast to int64 are now in the low bits of
	  the number, instead of the high bits.
	* Remove redundant talloc_free in LDAP.  Fixes #5927.
	* Add printing of prepend operator.  Fixes #5928.
	* Failed TCP / TLS connections no longer mark the home server as "dead".
	  Existing connections (if any) continue to be used.  New connection are
	  blocked for "connect_fail_interval" or "certificate_fail_interval".

FreeRADIUS 3.2.10 Wed 03 Jun 2026 12:00:00 UTC urgency=high
	Configuration changes

	Feature improvements

	Bug fixes
	* Correct bug where detail file reader would not read files.
	  Patch from Bjørn Mork.  Fixes #5870

FreeRADIUS 3.2.9 Mon 01 Jun 2026 12:00:00 UTC urgency=low
	Configuration changes
	* Add "protocol_error = yes" configuration to clients.  If set,
	  the server can return Protocol-Error responses to the client.
	* radclient can now suppress Message-Authenticator in Access-Request,
	  when the input packet contains "Message-Authenticator !* ANY"
	  Don't use this in production!
	* Set "suppress_secrets = true" by default.
	* Add "connect_fail_interval" to home_server configuration.  If a
	  connection fails, the server will wait this time before trying to
	  connect again.
	* Add "certificate_fail_interval" to home_server configuration.
	  If a connection succeeds but the home_server certificate is
	  invalid, the server will wait this time before trying to connect
	  again.
	* Add `update` section to home_server configuration.
	  Status-Server packets can therefore be customized.
	* Add "cipher_suites" to tls{} configuration.  See
	  raddb/sites-available/tls.  This is mainly used to set the
	  cipher suites for TLS-PSK with TLS 1.3.

	Feature improvements
	* Initial implementation of Protocol-Failure as per IETF draft.
	  The functionality is disabled by default, but can be enabled
	  via new configuration flags.
	* Always allow Protocol-Error packet as valid response to any
	  packet.
	* Add Error-Cause attributes to CoA-NAK and Disconnect-NAK
	* added filter_username_nai to policy.d/filter, mainly for use in
	  eduroam.
	* Updates to VSCode default configuration.
	* Cleanups and add log messages for rlm_proxy_rate_limit.
	* allow 389ds legacy PBKDF2_SHA256 to use arbitrary iteration
	  count.  (#5654)
	* amend policy insert_acct_class/acct_unique to work in
	  environments with multiple Class attributes (#5337)
	* Tweak sqlippool messages to make them clearer.
	* Print log message if the server receives a correct authenticated
	  proxy response packet, but which has an unexpected code.
	  e.g. received Access-Accept in response to an Accounting-Request.
	* New installations now set "suppress_secrets=true" by default.
	  The server also prints messages in debug mode which explains why
	  the secrets are being suppressed.
	* Allow parallel build for Debian.  Fixes #5774.
	* Add RTBrick and other dictionaries.
	* Add documentation for ntlm_auth and spaces in passwords.
	  Addresses #5654.

	Bug fixes
	* Many minor bug fixes and cleanups.
	* Fixes to RadSec.
	* Many other fixes to socket and event handling, which enable
	  increased scalability.
	* Fix issues found with EAP-MSCHAPv2, EAP-PWD, and EAP-MD5.
	* Fix run_dir (#5637) and MemoryLimit (#5639)
	* Disable the PCRE JIT at run time if it can't allocate
	  executable memory.
	* Set selinux boolean to allow PCRE2 JIT
	* If you set the clock 25 years in the future, don't spam
	  systemd.  Fixes #5642
	* Don't load the OpenSSL legacy provider when built with
	  --enable-fips-workaround.  Fixes #5644.
	* Address potential leaks when opening many RADIUS/TLS proxy
	  sockets.
	* Encode multiple DHCP Option 82 as one option, instead of as
	  multiple options.
	* Update the rlm_cache_redis driver to reconnect on connection
	  failure.  Fixes #5651.
	* Tweaks to the processing state machine to handle more corner

src/main/builddbm.c
/*
 * builddbm	Build a DBM file from an ASCII users file.
 *
 * Version:	@(#)builddbm  2.2  10-Aug-1999  miquels@cistron.nl
 *
 */
char sccsid[] =
"@(#)builddbm.c	2.2 Copyright 1999 Cistron Internet Services";

#include	"autoconf.h"

#include	<sys/types.h>
#include	<sys/stat.h>

#include	<stdio.h>
#include	<strings.h>
#include	<fcntl.h>
#include	<stdlib.h>
#include	<unistd.h>
#ifdef WITH_DBM
#  include	<dbm.h>
#endif
#ifdef WITH_NDBM
#  include	<ndbm.h>
#endif

#include	"radiusd.h"

const char	*progname;
int		debug_flag;
char		*radius_dir;
char		*radlog_dir;

FILE		*userfd;


/*
 *	Print a list of VALUE_PAIRS into a string,
 *	separated by comma's and closed off with a newline.
 */
void makelist(char *out, int outlen, VALUE_PAIR *vp)
{
	char		*ptr;
	int		len;

	ptr = out;
	ptr[0] = 0;

	while (vp && outlen > 3) {
		vp_prints(ptr, outlen, vp);
		strcat(ptr, ", ");
		len = strlen(ptr);
		outlen -= len + 2;
		ptr += len;
		vp = vp->next;
	}
	strcat(ptr, "\n");
}


int main(int argc, char **argv)
{
	PAIR_LIST	*users;
	char		name[MAX_STRING_LEN];
	char		content[4096];
	int		len;
	datum		named;
	datum		contentd;
	int		defno = 0;
#ifdef WITH_DBM
	int		fd;
#endif
#ifdef WITH_NDBM
	DBM	*dbm;
#endif

	progname = argv[0];
	radius_dir = ".";
	librad_dodns = 0;

	if (dict_init(RADDBDIR, RADIUS_DICTIONARY) < 0) {
		librad_perror("builddbm");
		return 1;
	}

	/*
	 *	Read the "users" file.
	 */
	if ((users = pairlist_read(RADIUS_USERS, 1)) == NULL)
		exit(1);

	/*
	 *	Initialize a new, empty database.
	 */
	umask(077);
#ifdef WITH_DBM
	if ((fd = open("users.pag", O_WRONLY | O_CREAT | O_TRUNC, 0600)) < 0) {
		fprintf(stderr, "%s: Couldn't open users.pag for writing\n",
			progname);
		exit(1);
	}
	close(fd);
	if ((fd = open("users.dir", O_WRONLY | O_CREAT | O_TRUNC, 0600)) < 0) {
		fprintf(stderr, "%s: Couldn't open users.dir for writing\n",
			progname);
		exit(1);
	}
	close(fd);
	if (dbminit("users") != 0) {
		fprintf(stderr, "%s: ", progname);
		perror("dbminit(users)");
		exit(1);
	}
#endif
#ifdef WITH_NDBM
	if ((dbm = dbm_open("users", O_RDWR|O_CREAT|O_TRUNC, 0600)) == NULL) {
		fprintf(stderr, "%s: ", progname);
		perror("dbm_open(users)");
		exit(1);
	}
#endif

	while (users) {

		makelist(content, sizeof(content), users->check);
		len = strlen(content);
		makelist(content + len, sizeof(content) - len, users->reply);

		strNcpy(name, users->name, sizeof(name));
		if (strcmp(name, "DEFAULT") == 0) {
			if (defno > 0)
				sprintf(name, "DEFAULT%d", defno);
			defno++;
		}
		named.dptr = name;
		named.dsize = strlen(name);
		contentd.dptr = content;
		contentd.dsize = strlen(content);
#ifdef WITH_DBM
		if (store(named, contentd) != 0)
#endif
#ifdef WITH_NDBM
		if (dbm_store(dbm, named, contentd, DBM_INSERT) != 0)
#endif
		{
			fprintf(stderr, "%s: Couldn't store datum for %s\n",
				progname, name);
			exit(1);
		}
		users = users->next;
	}
#ifdef WITH_DBM
	dbmclose();
#endif
#ifdef WITH_NDBM
	dbm_close(dbm);
#endif
	return 0;
}

